// module_gm/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/*
 * @Author: zhaolei
 * @Date: 2022-05-05 09:57:23
 * @LastEditors: zhaolei
 * @Description: file content
 */
/**
 * 下拉列表元素
 */
class DropdownItem extends UIScript {
    /** 下拉列表 */
    list;
    /** 数据 */
    data;
}
/**
 * 下拉列表
 */
class DropdownList {
    _root;
    _itemCls;
    space;
    /** 不显示的缓存道具 */
    _cache;
    /** 显示的道具 */
    _items;
    /** 是否下拉 */
    _isDropdown;
    /** 点击委托 */
    onSelect = new Action1();
    constructor(_root, _itemCls, space = 0.5) {
        this._root = _root;
        this._itemCls = _itemCls;
        this.space = space;
        /** 缓存 */
        this._cache = [];
        /** 所有条目 */
        this._items = [];
    }
    /**
    * 添加展开按钮事件
    */
    toggle() {
        this._isDropdown = !this._isDropdown;
        this._invalidateLayout();
    }
    /**
     * 添加展开按钮事件
     */
    expand() {
        this._isDropdown = true;
        this._invalidateLayout();
    }
    /**
     * 添加展开按钮事件
     */
    unExpand() {
        this._isDropdown = false;
        this._invalidateLayout();
    }
    /**
     * 添加一个选项
     * @param node
     * @param index 索引
     */
    addItem(data, index = -1) {
        let itemUI = this._cache.length > 0 ? this._cache.shift() : UIService.create(this._itemCls);
        if (!itemUI.list) {
            itemUI.list = this;
            itemUI.button.touchMethod = ButtonTouchMethod.PreciseTap;
            itemUI.button.onClicked.add(() => {
                this.onSelect.call(itemUI.button.text);
                this.unExpand();
            });
            this._root.cmdPanel.addChild(itemUI.uiObject);
        }
        itemUI.uiObject.visibility = SlateVisibility.SelfHitTestInvisible;
        itemUI.button.text = data;
        if (index >= 0) {
            this._items.splice(index, 0, itemUI);
        }
        else {
            this._items.push(itemUI);
        }
        this._root.cmdPanel.size = new Vector2(270, this._items.length * (itemUI.uiObject.size.y + this.space));
        this._invalidateLayout();
    }
    /**
     * 删除一个选项
     * @param node
     */
    removeItem(node) {
        const index = this._items.indexOf(node);
        if (index >= 0) {
            node.visible = false;
            this._cache.push(...this._items.splice(index, 1));
            this._invalidateLayout();
        }
    }
    /**
     * 清楚列表
     */
    clear() {
        for (const item of this._items) {
            if (!this._cache.includes(item)) {
                this._cache.push(item);
            }
            item.uiObject.visibility = SlateVisibility.Collapsed;
        }
        this._items.length = 0;
        this._invalidateLayout();
    }
    /**
     * 删除一个指定索引
     * @param index
     */
    removeItemAt(index) {
        const node = this.getItem(index);
        if (node) {
            this.removeItem(node);
        }
    }
    /**
     * 获取一个选项,超出范围则返回空
     * @param index
     */
    getItem(index) {
        if (index >= 0 && index < this._items.length)
            return this._items[index];
        return null;
    }
    /**
     * 获取选项数量
     */
    get size() {
        return this._items.length;
    }
    /**
     * 重新对齐面板
     */
    _invalidateLayout() {
        if (this._isDropdown) {
            let offset = 0;
            this._root.cmdPanel.position = Vector2.zero;
            this._root.dropList.visibility = SlateVisibility.SelfHitTestInvisible;
            for (let i = 0; i < this._items.length; i++) {
                this._items[i].uiObject.position = new Vector2(0, offset);
                offset += this._items[i].uiObject.size.y + this.space;
            }
        }
        else {
            this._root.dropList.visibility = SlateVisibility.Collapsed;
        }
    }
}

/**
 * @Author       : lei.zhao
 * @Date         : 2022-12-19 18:33:40
 * @LastEditors  : lei.zhao
 * @LastEditTime : 2023-04-10 15:50:21
 * @FilePath     : \commonmodule_gm\JavaScripts\module\gmModule\GMConfig.ts
 * @Description  : 修改描述
 */
/** GM配置 */
const GMConfig = [];
/**
 * 添加GM命令
 * @param label 名称
 * @param cmdClient 客户端命令
 * @param cmdServer 服务端命令
 * @param shortKey 快捷键
 */
function AddGMCommand(label, cmdClient, cmdServer, group, shortKey) {
    if (shortKey != null) {
        label = label + "(" + shortKey + ")";
    }
    const data = { label: label, group: group, clientCmd: cmdClient, serverCmd: cmdServer, shortKey: shortKey };
    GMService.instance.init();
    if (SystemUtil.isClient()) {
        if (shortKey != null) {
            InputUtil.onKeyDown(shortKey, () => {
                GMService.instance.cmd(data, exports.GM.getParam ? exports.GM.getParam() : "");
            });
        }
    }
    GMConfig.push(data);
}

/**
 * @Author       : lei.zhao
 * @Date         : 2022-12-19 18:33:40
 * @LastEditors  : xiaohao.li
 * @LastEditTime : 2023-11-16 09:22:31
 * @FilePath     : \commonmodule_gm\JavaScripts\module\gmModule\GMService.ts
 * @Description  : 修改描述
 */
/**
 * GM服务, 用于初始化GM配置，调用GM命令
 */
class GMService {
    /** GM服务单例 */
    static _instance;
    /** 是否获取了GM用户列表 */
    isGetGMUserList = false;
    /** 有权限的用户列表 */
    userList = [];
    /** 是否有GM权限的回调 */
    authorityCallback;
    /** 是否已经初始化 */
    isInited = false;
    /**
     * 获取GM服务单例
     * @returns GM服务单例
     */
    static get instance() {
        if (!GMService._instance) {
            GMService._instance = new GMService();
        }
        return GMService._instance;
    }
    /**
     * 初始化GM服务
     */
    init() {
        if (this.isInited)
            return;
        this.isInited = true;
        if (SystemUtil.isServer()) {
            mw.Event.addClientListener("GM_CALL", (player, gmIndex, param) => {
                GMConfig[gmIndex].serverCmd(player, param);
            });
            mw.Event.addClientListener("GM.CheckAuthority", (player, isTemporaryStorage) => {
                this.getAuthority(isTemporaryStorage, player.userId).then((isAuthority) => {
                    mw.Event.dispatchToClient(player, "GM.CheckAuthority", isAuthority);
                });
            });
        }
        else {
            mw.Event.addServerListener("GM.CheckAuthority", (isAuthority) => {
                this.authorityCallback && this.authorityCallback(isAuthority);
            });
        }
    }
    /**
     * 初始化UI
     */
    createUI(dropDownList) {
        GMConfig.forEach(cmd => {
            dropDownList.addItem(cmd);
        });
    }
    constructor() {
    }
    /**
     * 获取权限
     * @param isTemporaryStorage 是否临时存储
     * @param userId 用户ID
     */
    async getAuthority(isTemporaryStorage, userId) {
        if (!this.isGetGMUserList) {
            this.isGetGMUserList = true;
            DataStorage.setTemporaryStorage(isTemporaryStorage);
            const result = await DataStorage.asyncGetData("GM.UserList");
            if (result.code === DataStorageResultCode.Success) {
                const userList = result.data;
                if (!userList) {
                    DataStorage.asyncSetData("GM.UserList", []);
                }
                else {
                    this.userList = userList;
                }
            }
        }
        return this.userList.includes(userId);
    }
    /**
     * 客户调用，判断主控端是否有GM权限
     * @param isTemporaryStorage 是否临时存储
     * @param callback 回调
     */
    checkGMAuthority(isTemporaryStorage, callback) {
        if (SystemUtil.isPIE) {
            callback(true);
            return;
        }
        mw.Event.dispatchToServer("GM.CheckAuthority", isTemporaryStorage);
        this.authorityCallback = callback;
    }
    /**
     * 调用命令
     * @param data
     */
    cmd(data, param) {
        if (data.clientCmd) {
            data.clientCmd(Player.localPlayer, param);
        }
        mw.Event.dispatchToLocal("GM_RUN", data.label);
        const index = GMConfig.indexOf(data);
        if (index >= 0 && data.serverCmd) {
            mw.Event.dispatchToServer("GM_CALL", index, param);
        }
    }
}

/**
 * @Author       : lei.zhao
 * @Date         : 2023-03-19 13:31:37
 * @LastEditors  : lei.zhao
 * @LastEditTime : 2023-07-23 14:42:46
 * @FilePath     : \commonmodule_gm\JavaScripts\module\gmModule\GM.ts
 * @Description  : 修改描述
 */
/**
 * GM模块
 */
exports.GM = void 0;
(function (GM) {
    /**
     * 客户端调用，开始GM
     * @param uiClass
     */
    function start(uiClass) {
        if (SystemUtil.isClient()) {
            new uiClass().show();
        }
    }
    GM.start = start;
    /**
     * 客户端调用，查看是否有GM权限
     * ```
     * 权限来自开发者后台自定义存档GM.UserList
     * 结构为["userId1","userId2","userId3",....]
     * 如需增加GM权限，只需在存档中添加userId即可
     * ```
     * @param callback true-有GM权限，false-无GM权限
     * @param isTemporaryStorage 是否是临时存档 default: SystemUtil.isPIE
     */
    function checkAuthority(callback, isTemporaryStorage = SystemUtil.isPIE) {
        GMService.instance.checkGMAuthority(isTemporaryStorage, callback);
    }
    GM.checkAuthority = checkAuthority;
})(exports.GM || (exports.GM = {}));

/**
 * @Author       : lei.zhao
 * @Date         : 2022-12-19 18:33:40
 * @LastEditors  : xiaohao.li
 * @LastEditTime : 2023-11-16 09:31:23
 * @FilePath     : \commonmodule_gm\JavaScripts\module\gmModule\GMBasePanel.ts
 * @Description  : 修改描述
 */
/**
 * GM基础面板
 */
class GMBasePanel {
    /** 下拉列表 */
    dropDownList;
    /** 是否是分组 */
    isGroup = false;
    /** 分组名称 */
    groupText = "分组";
    /** 命令名称 */
    cmdText = "命令";
    /** UI实例 */
    _view;
    constructor(_viewCls, _itemCls) {
        exports.GM.getParam = () => {
            return this._view.argText.text;
        };
        this._view = UIService.create(_viewCls);
        this._view.groupButton.onClicked.add(() => {
            this.dropDownList.clear();
            this.dropDownList.addItem("分组");
            const groupList = [];
            GMConfig.forEach(cmd => {
                if (cmd.group && !groupList.includes(cmd.group)) {
                    groupList.push(cmd.group);
                    this.dropDownList.addItem(cmd.group);
                }
            });
            this.dropDownList.toggle();
            this.isGroup = true;
        });
        this._view.cmdButton.onClicked.add(() => {
            this.dropDownList.clear();
            if (this.groupText == "分组") {
                GMConfig.forEach(cmd => {
                    this.dropDownList.addItem(cmd.label);
                });
            }
            else {
                GMConfig.forEach(cmd => {
                    if (cmd.group == this.groupText) {
                        this.dropDownList.addItem(cmd.label);
                    }
                });
            }
            this.dropDownList.toggle();
            this.isGroup = false;
        });
        this._view.okButton.onClicked.add(() => {
            const cmd = GMConfig.find(i => i.label == this.cmdText);
            cmd && GMService.instance.cmd(cmd, this._view.argText.text);
        });
        this.dropDownList = new DropdownList(this._view, _itemCls, 1);
        this.dropDownList.onSelect.add(text => {
            this.isGroup ? (this.groupText = text, this._view.groupButton.text = text) : (this.cmdText = text, this._view.cmdButton.text = text);
        });
        mw.Event.addLocalListener("GM_RUN", (label) => {
            this._view.cmdButton.text = label;
        });
        GMService.instance.createUI(this.dropDownList);
    }
    /**
     * 显示GM面板
     */
    show() {
        UIService.showUI(this._view);
    }
}

exports.AddGMCommand = AddGMCommand;
exports.DropdownItem = DropdownItem;
exports.DropdownList = DropdownList;
exports.GMBasePanel = GMBasePanel;
exports.GMConfig = GMConfig;
exports.GMService = GMService;
