// mw-god-mod/dist/ui/GodModPanel.js
import GodModStringParamInput from './param-input/GodModStringParamInput.js';
import GodModNumberParamInput from './param-input/GodModNumberParamInput.js';
import GodModIntegerParamInput from './param-input/GodModIntegerParamInput.js';
import GodModVectorParamInput from './param-input/GodModVectorParamInput.js';
import Gtk from 'gtoolkit';
import { Lui, Component, Button, Property, AutoComplete } from 'mw-lynx-ui';
import { GodModPanelSizeX } from './base/GodModPanelConst.js';
import GodModEnumParamInput from './param-input/GodModEnumParamInput.js';

var Color = Lui.Asset.Color;
var ColorUtil = Lui.Asset.ColorUtil;
var Interval = Lui.Asset.Interval;
class GodModPanel extends Component {
    constructor() {
        super(...arguments);
        this._godCommandItems = [];
        this._paramCache = new Map();
        this._paramInputComponentCache = new Map();
        this._lastShowTipsTime = 0;
        this.renderAnimHandler = dt => {
            if (this._txtInfo.renderOpacity > 0) {
                const d = Date.now() - this._lastShowTipsTime;
                if (d > GodModPanel.TipsShownTime) {
                    this._txtInfo.renderOpacity = Math.max(0, this._txtInfo.renderOpacity - dt / (GodModPanel.TipsFadeTime / 1000));
                }
            }
            const currentY = this._cnvParamInput.position.y;
            const shrinkSize = this._cnvParamInput.size.y - this._txtInfo.size.y;
            if (this._currentChoose !== undefined && currentY < 0) {
                Gtk.setUiPositionY(this._cnvParamInput, Math.min(0, currentY + shrinkSize * dt / Interval.Fast));
            }
            if (this._currentChoose === undefined) {
                if (currentY > -shrinkSize) {
                    Gtk.setUiPositionY(this._cnvParamInput, Math.max(-shrinkSize, currentY - shrinkSize * dt / Interval.Fast));
                }
            }
        };
        this.handleDrag = () => {
            if (this._dragStartTime === undefined)
                return;
            if (Date.now() - this._dragStartTime < this._dragSensitive)
                return;
            const viewPortCanvas = mw.UIService.canvas;
            if (this._mouseStartMosPos === undefined) {
                this._mouseStartMosPos = mw.absoluteToLocal(viewPortCanvas.cachedGeometry, mw.getMousePositionOnPlatform());
                this._mouseStartCnvPos = this.root.position;
                this.playStartDragEffect();
            }
            else {
                let currMouseRelativePos = mw.absoluteToLocal(viewPortCanvas.cachedGeometry, mw.getMousePositionOnPlatform());
                Gtk.setUiPosition(this.root, this._mouseStartCnvPos.x + currMouseRelativePos.x - this._mouseStartMosPos.x, this._mouseStartCnvPos.y + currMouseRelativePos.y - this._mouseStartMosPos.y);
            }
        };
        //#region CallBack
        //#endregion ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
    }
    //#region Lui Component
    static create(option) {
        let godModPanel = new GodModPanel();
        godModPanel._godCommandItems = option?.items ?? [];
        if (option.zOrder !== undefined)
            godModPanel.root.zOrder = option.zOrder;
        godModPanel._dragSensitive = option?.dragSensitive ?? 0.5e3;
        Gtk.setUiSize(godModPanel.root, GodModPanelSizeX, 140);
        godModPanel._cnvController = mw.Canvas.newObject(godModPanel.root, "cnvController");
        Gtk.setUiSize(godModPanel._cnvController, GodModPanelSizeX, 80);
        Gtk.trySetVisibility(godModPanel._cnvController, true);
        godModPanel._btnExpand = Button.create({
            label: "expand",
            size: { x: 220, y: 80 },
            padding: { top: 10, right: 20, bottom: 10, left: 0 },
            color: {
                primary: Color.Blue,
                secondary: Color.Blue200,
            },
            fontSize: 16,
            corner: Property.Corner.TopRight | Property.Corner.Bottom,
        }).attach(godModPanel._cnvController);
        Gtk.setUiPosition(godModPanel._btnExpand.root, 0, 0);
        godModPanel._btnExpand.onClick.add(() => godModPanel.showTips("打破常规的全新呼出方式. 敬请期待..."));
        godModPanel._btnMove = Button.create({
            label: "move",
            size: { x: 100, y: 80 },
            padding: { top: 10, right: 20, bottom: 10, left: 0 },
            color: {
                primary: Color.Blue,
                secondary: Color.Blue200,
            },
            fontSize: 16,
            corner: Property.Corner.All,
        }).attach(godModPanel._cnvController);
        Gtk.setUiPosition(godModPanel._btnMove.root, godModPanel._btnExpand.root.size.x, 0);
        godModPanel._btnMove.onPress.add(() => {
            godModPanel._dragStartTime = Date.now();
        });
        godModPanel._btnMove.onRelease.add(() => {
            godModPanel.playStopDragEffect();
            godModPanel._dragStartTime = undefined;
            godModPanel._mouseStartMosPos = undefined;
            godModPanel._mouseStartCnvPos = undefined;
        });
        godModPanel._btnClose = Button.create({
            label: "close",
            size: { x: 80, y: 80 },
            padding: { top: 10, right: 0, bottom: 10, left: 0 },
            color: {
                primary: Color.Red,
                secondary: Color.Red200,
            },
            fontSize: 16,
            icon: "287315",
            corner: Property.Corner.TopLeft | Property.Corner.Bottom,
        }).attach(godModPanel._cnvController);
        Gtk.setUiPosition(godModPanel._btnClose.root, godModPanel._btnExpand.root.size.x + godModPanel._btnMove.root.size.x, 0);
        godModPanel._btnClose.onClick.add(godModPanel.detach, undefined, undefined, godModPanel);
        godModPanel._acInput = AutoComplete.create({
            label: "search commands",
            items: godModPanel._godCommandItems,
            size: { x: GodModPanelSizeX, y: 60 },
            color: {
                primary: Color.Blue,
                secondary: Color.Blue200,
            },
            itemHeight: 40,
            maxCount: 6,
            fontSize: 16,
            fontStyle: mw.UIFontGlyph.Light,
            variant: "filled",
            additionKey: [
                { name: "pinyin", getFn: (item) => item.pinyin },
            ],
            // renderOption: (item) => {},
            corner: Property.Corner.Top,
            zOrder: 5,
        })
            .attach(godModPanel);
        Gtk.setUiPosition(godModPanel._acInput.root, 0, godModPanel._cnvController.size.y);
        godModPanel._cnvParamInputContainer = mw.Canvas.newObject(godModPanel.root, "cnvParamInputContainer");
        Gtk.setUiPosition(godModPanel._cnvParamInputContainer, 0, 140);
        Gtk.setUiSize(godModPanel._cnvParamInputContainer, GodModPanelSizeX, 1080);
        Gtk.trySetVisibility(godModPanel._cnvParamInputContainer, true);
        godModPanel._cnvParamInputContainer.clipEnable = true;
        godModPanel._cnvParamInput = mw.Canvas.newObject(godModPanel._cnvParamInputContainer, "cnvParamInputContainer");
        Gtk.setUiPosition(godModPanel._cnvParamInput, 0, 0);
        godModPanel._acInput.onClear.add(() => {
            godModPanel.hideCnvParamInput();
        });
        godModPanel._acInput.onChoose.add(event => {
            godModPanel._currentChoose = event.item;
            godModPanel.showCnvParamInput();
        });
        godModPanel._btnRun = Button.create({
            label: "Run",
            size: { x: GodModPanelSizeX, y: this.BtnRunSizeY },
            color: {
                primary: Color.Green,
                secondary: Color.Green200,
            },
            fontSize: 24,
            corner: Property.Corner.Top,
        }).attach(godModPanel._cnvParamInput);
        godModPanel._btnRun.onClick.add(_ => godModPanel.commit());
        godModPanel._txtInfo = mw.TextBlock.newObject(godModPanel._cnvParamInput, "txtInfo");
        Gtk.setUiSize(godModPanel._txtInfo, GodModPanelSizeX, this.TxtInfoSizeY);
        godModPanel._txtInfo.fontSize = 16;
        godModPanel._txtInfo.textAlign = mw.TextJustify.Left;
        godModPanel._txtInfo.textHorizontalLayout = mw.UITextHorizontalLayout.NoClipping;
        godModPanel._txtInfo.renderOpacity = 0;
        godModPanel._txtInfo.setOutlineColorByHex(ColorUtil.colorHexWithAlpha(Color.Gray800, 1));
        godModPanel._txtInfo.outlineSize = 2;
        godModPanel._imgDrag = mw.Image.newObject(godModPanel.root, "imgDrag");
        Gtk.setUiSize(godModPanel._imgDrag, GodModPanelSizeX, 150);
        godModPanel._imgDrag.imageGuid = Lui.Asset.ImgRoundedRectangle;
        godModPanel._imgDrag.imageDrawType = mw.SlateBrushDrawType.PixcelBox;
        godModPanel._imgDrag.margin = new mw.Margin(Lui.Asset.ImgRoundedRectangleBoxMargin.left, Lui.Asset.ImgRoundedRectangleBoxMargin.top, Lui.Asset.ImgRoundedRectangleBoxMargin.right, Lui.Asset.ImgRoundedRectangleBoxMargin.bottom);
        godModPanel._imgDrag.setImageColorByHex(ColorUtil.colorHexWithAlpha(Color.White, 0.5));
        Gtk.trySetVisibility(godModPanel._imgDrag, false);
        godModPanel.onAttach.add(() => {
            mw.TimeUtil.onEnterFrame.add(godModPanel.handleDrag);
        });
        godModPanel.onDetach.add(() => {
            mw.TimeUtil.onEnterFrame.remove(godModPanel.handleDrag);
        });
        return godModPanel;
    }
    static defaultOption() {
        return undefined;
    }
    destroy() {
        super.destroy();
    }
    //#endregion ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
    registerCommandHandler(handler) {
        this._runCommandHandler = handler;
        return this;
    }
    //#region Init
    //#endregion ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
    showCnvParamInput() {
        if (this._lastChoose === this._currentChoose)
            return;
        if (this._lastChoose) {
            this._paramCache.set(this._lastChoose, this._currentInputComponent?.getParam() ?? undefined);
        }
        if (!this._btnRun.enable)
            this._btnRun.enable = true;
        this._lastChoose = this._currentChoose;
        const type = this._currentChoose.paramType;
        this._currentInputComponent?.detach();
        this._btnRun.detach();
        this._txtInfo.removeObject();
        let input;
        if (type !== "void") {
            input = this._paramInputComponentCache.get(type);
            if (!input) {
                switch (type) {
                    case "integer":
                        input = GodModIntegerParamInput.create();
                        break;
                    case "number":
                        input = GodModNumberParamInput.create();
                        break;
                    case "vector":
                        input = GodModVectorParamInput.create();
                        break;
                    case "string":
                    default:
                        if (typeof type === "object") {
                            input = GodModEnumParamInput.create({ enumObj: type });
                        }
                        else {
                            input = GodModStringParamInput.create();
                        }
                        break;
                }
                input.onCommit.add(() => {
                    if (this._currentInputComponent === input) {
                        this._btnRun.enable = input.validated.result;
                    }
                });
                this._paramInputComponentCache.set(type, input);
            }
            if (input) {
                input.setValidator(this._currentChoose.paramOption?.validator);
                input.setParam(this._paramCache.get(this._currentChoose));
            }
            if (typeof type === "object") {
                input.setEnumObj(type);
            }
        }
        const paramSizeY = (input?.root?.size.y ?? 0);
        const paramAreaSizeY = paramSizeY + GodModPanel.BtnRunSizeY + GodModPanel.TxtInfoSizeY;
        Gtk.setUiSizeY(this._cnvParamInput, paramAreaSizeY);
        Gtk.setUiPositionY(this._btnRun.root, paramSizeY);
        Gtk.setUiPositionY(this._txtInfo, paramSizeY +
            this._btnRun.root.size.y);
        this._currentInputComponent = input?.attach(this._cnvParamInput) ?? undefined;
        this._btnRun.attach(this._cnvParamInput);
        this._cnvParamInput.addChild(this._txtInfo);
    }
    hideCnvParamInput() {
        if (this._lastChoose) {
            this._paramCache.set(this._lastChoose, this._currentInputComponent?.getParam() ?? undefined);
        }
        this._currentChoose = undefined;
    }
    commit() {
        if (this._currentInputComponent && !this._currentInputComponent.validated.result) {
            this.showWarning(this._currentInputComponent.validated.reason);
            this._btnRun.enable = false;
            return;
        }
        const param = this._currentInputComponent?.getParam() ?? undefined;
        let command = this._currentChoose;
        this._runCommandHandler?.(command.label, param);
        this.showSuccess();
    }
    playStartDragEffect() {
        Gtk.setUiScale(this.root, 1.1, 1.1);
        Gtk.trySetVisibility(this._imgDrag, true);
    }
    playStopDragEffect() {
        Gtk.setUiScale(this.root, 1, 1);
        Gtk.trySetVisibility(this._imgDrag, false);
        if (this.root.position.x < 0) {
            Gtk.setUiPositionX(this.root, 0);
        }
        if (this.root.position.y < 0) {
            Gtk.setUiPositionY(this.root, 0);
        }
        if (this.root.position.x > mw.UIService.canvas.size.x - this.root.size.x) {
            Gtk.setUiPositionX(this.root, mw.UIService.canvas.size.x - this.root.size.x);
        }
        if (this.root.position.y > mw.UIService.canvas.size.y - this.root.size.y) {
            Gtk.setUiPositionY(this.root, mw.UIService.canvas.size.y - this.root.size.y);
        }
    }
    showTips(text, color) {
        this._txtInfo.setFontColorByHex(ColorUtil.colorHexWithAlpha(color ?? Color.Blue, 1));
        this._txtInfo.text = text;
        this._lastShowTipsTime = Date.now();
        this._txtInfo.renderOpacity = 1;
    }
    showWarning(text) {
        this.showTips(text, Color.Red);
    }
    showSuccess() {
        this.showTips("Worked!", Color.Green);
    }
}
//#region Constant
GodModPanel.TipsShownTime = 2e3;
GodModPanel.TipsFadeTime = 1e3;
GodModPanel.BtnRunSizeY = 60;
GodModPanel.TxtInfoSizeY = 30;

export { GodModPanel };
