// mw-god-mod/dist/GodModService.js
export { RangeDataValidator } from './GodModParam.js';
import { GodCommandItem } from './GodCommandItem.js';
import { GodModPanel } from './ui/GodModPanel.js';
export { GodModParamInputBase } from './ui/param-base/GodModParamInputBase.js';
export { ParamInputSizeY, ParamInputZOrder } from './ui/param-base/IGodModParamInput.js';
import './ui/param-input/GodModIntegerParamInput.js';
import './ui/param-input/GodModNumberParamInput.js';
import './ui/param-input/GodModStringParamInput.js';
import './ui/param-input/GodModVectorParamInput.js';
export { ExpandIcon } from './ui/icon/ExpandIcon.js';
export { MoveIcon } from './ui/icon/MoveIcon.js';
import Gtk, { Singleton, Regulator, GtkTypes } from 'gtoolkit';
import Log4Ts from 'mw-log4ts';

/**
 * God Mod 服务.
 *
 * ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟
 * ⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄
 * ⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄
 * ⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄
 * ⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
 * @author LviatYi
 * @font JetBrainsMono Nerd Font Mono https://github.com/ryanoasis/nerd-fonts/releases/download/v3.0.2/JetBrainsMono.zip
 * @fallbackFont Sarasa Mono SC https://github.com/be5invis/Sarasa-Gothic/releases/download/v0.41.6/sarasa-gothic-ttf-0.41.6.7z
 */
class GodModService extends Singleton() {
    constructor() {
        super(...arguments);
        //#endregion ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
        this._queriedAdminList = false;
        /**
         * GodCommands 库.
         * @type {Map<string, GodCommandItem<unknown>>}
         * @private
         */
        this._commands = new Map();
        /**
         * 管理员列表.
         * @type {Set<string>}
         * @private
         */
        this._adminList = undefined;
        this._shutdown = false;
        this._checkAuthorityRegulator = new Regulator(GtkTypes.Interval.PerMin * 5);
    }
    onConstruct() {
        if (mw.SystemUtil.isClient()) {
            mw.Event.addServerListener(GodModService.GodModCommandRunResultInServerEventName, (event) => {
                const e = event;
                if (this._currentFrontFocus === e.label) {
                    this.showUiResult(e.result);
                }
            });
        }
        if (mw.SystemUtil.isServer()) {
            mw.TimeUtil.onEnterFrame.add(() => {
                if (this._checkAuthorityRegulator.request()) {
                    this.queryAdminList();
                }
            });
            mw.Event.addClientListener(GodModService.GodModQueryAuthorityReqEventName, player => {
                mw.Event.dispatchToClient(player, GodModService.GodModQueryAuthorityRespEventName, this.verifyStrongAuthority(player.userId));
            });
        }
    }
    addCommand(label, paramType = "string", clientCmd = undefined, serverCmd = undefined, paramOption = undefined, group) {
        if (this._shutdown)
            return;
        if (this._commands.get(label)) {
            Log4Ts.error(GodModService, `A command with the same label already exists.`, label);
        }
        else if (!clientCmd && !serverCmd) {
            Log4Ts.error(GodModService, `at least one of the client command and server command must be provided.`);
        }
        else {
            this._commands.set(label, new GodCommandItem(label, paramType, clientCmd, serverCmd, paramOption, group));
            if (mw.SystemUtil.isServer()) {
                mw.Event.addClientListener(this.getEventName(label), (player, p) => {
                    this.runCommandInServer(player, label, p);
                });
            }
        }
    }
    removeCommand(label) {
        this._commands.delete(label);
    }
    getAllCommands() {
        return [...this._commands.values()];
    }
    runCommandInClient(label, p, autoDispatchToServer = true) {
        if (this._shutdown || !mw.SystemUtil.isClient())
            return;
        const command = this._commands.get(label);
        if (!command || !command.isParamValid(p))
            return;
        this._currentFrontFocus = label;
        Log4Ts.log(GodModService, `run command in client.`, `command: ${this._currentFrontFocus}`);
        let result = false;
        try {
            let r = command.clientCmd?.(p);
            if (r !== false)
                result = true;
        }
        catch (e) {
            Log4Ts.error(GodModService, `error occurs in client command.`, e);
        }
        if (command.serverCmd && autoDispatchToServer) {
            Event.dispatchToServer(this.getEventName(command.label), p);
        }
        else {
            this.showUiResult(result);
        }
    }
    runCommandInServer(player, label, p) {
        if (this._shutdown || !mw.SystemUtil.isServer())
            return;
        if (!this.verifyWeakAuthority(player.userId))
            return;
        const command = this._commands.get(label);
        if (!command || !command.isParamValid(p))
            return;
        Log4Ts.log(GodModService, `run command in server.`, `command: ${this._currentFrontFocus}`, `by user: ${player.userId}`);
        let result = false;
        try {
            let r = command.serverCmd?.(player, p);
            if (r !== false)
                result = true;
            Event.dispatchToClient(player, GodModService.GodModCommandRunResultInServerEventName, { label: label, result });
        }
        catch (e) {
            Log4Ts.error(GodModService, `error occurs in server command.`, e);
            Event.dispatchToClient(player, GodModService.GodModCommandRunResultInServerEventName, { label: label, result: false });
        }
    }
    /**
     * 显示 God Mod 面板.
     */
    showGm() {
        if (mw.SystemUtil.isClient()) {
            if (!this._view) {
                this._view = GodModPanel
                    .create({
                    items: Array.from(this._commands.values()),
                    zOrder: 650000,
                })
                    .attach(mw.UIService.canvas)
                    .registerCommandHandler((label, p, autoDispatchToServer) => {
                    this.runCommandInClient(label, p, autoDispatchToServer);
                });
            }
            else {
                this._view.attach(mw.UIService.canvas);
            }
        }
    }
    /**
     * 隐藏 God Mod 面板.
     */
    hideGm() {
        if (mw.SystemUtil.isClient()) {
            this._view?.detach();
        }
    }
    /**
     * 带强权限验证地 显示 God Mod 面板.
     * @desc 强权限验证意味着 必须手动添加到管理员列表.
     * @desc 当 PIE 环境时 无论如何将直接显示.
     */
    authShowGm() {
        if (mw.SystemUtil.isPIE)
            this.showGm();
        if (mw.SystemUtil.isClient()) {
            this.hasAuthority()
                .then(value => {
                if (value)
                    this.showGm();
                else
                    this.hideGm();
            });
        }
    }
    /**
     * 关闭 God Mod 服务.
     * @desc 执行后不再生效.
     */
    shutdown() {
        this._shutdown = true;
        this._commands.clear();
        return this;
    }
    //#region UI Config
    /**
     * 设置 GodMod 面板 zOrder.
     * @param {number} zOrder
     * @return {this}
     */
    setZOrder(zOrder) {
        if (this._view && this._view.root) {
            this._view.root.zOrder = zOrder;
        }
        return this;
    }
    /**
     * 设置 GodMod 面板 位置.
     * @param {number} x
     * @param {number} y
     * @return {this}
     */
    setPosition(x, y) {
        if (this._view && this._view.root) {
            Gtk.setUiPosition(this._view.root, x, y);
        }
        return this;
    }
    //#endregion ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
    /**
     * 是否 具有弱认证权限.
     * @desc 仅服务端.
     * @param {string} userId
     * @return {boolean}
     * @private
     */
    verifyWeakAuthority(userId) {
        return this._adminList === undefined || (this._adminList && this._adminList.has(userId));
    }
    /**
     * 是否 具有强认证权限.
     * @desc 仅服务端.
     * @param {string} userId
     * @return {boolean}
     * @private
     */
    verifyStrongAuthority(userId) {
        return this._adminList?.has(userId) ?? false;
    }
    /**
     * 是否 自身具有强认证权限.
     * @desc 仅客户端.
     * @desc 需服务器手动加入后生效.
     * @return {Promise<boolean>}
     */
    async hasAuthority() {
        if (!mw.SystemUtil.isClient()) {
            return false;
        }
        return new Promise((resolve, reject) => {
            const listener = mw.Event.addServerListener(GodModService.GodModQueryAuthorityRespEventName, (result) => {
                listener.disconnect();
                resolve(result);
            });
            mw.setTimeout(() => {
                listener.disconnect();
                reject();
            }, GtkTypes.Interval.PerSec * 3);
            mw.Event.dispatchToServer(GodModService.GodModQueryAuthorityReqEventName);
        });
    }
    /**
     * 查询管理员列表.
     */
    queryAdminList() {
        mw.DataStorage
            ?.asyncGetData(GodModService.GodModAdminListStorageKey)
            ?.then(result => {
            if (this._queriedAdminList)
                return;
            switch (result.code) {
                case mw.DataStorageResultCode.Success:
                    if (result.data === undefined) {
                        mw.DataStorage.asyncSetData(GodModService.GodModAdminListStorageKey, []);
                    }
                    else {
                        if (result.data?.length > 0 ?? false) {
                            this._adminList = new Set();
                            for (const d of result.data) {
                                this._adminList.add(d);
                            }
                        }
                        else {
                            this._adminList = undefined;
                        }
                    }
                    break;
            }
            this._queriedAdminList = true;
        });
    }
    ;
    /**
     * 获取填充类型名.
     * @param {string} label
     * @return {string}
     * @private
     */
    getEventName(label) {
        return GodModService.GodModRunRequestEventNamePrefix + label;
    }
    /**
     * 显示执行反馈结果。
     * @param {boolean} result
     * @private
     */
    showUiResult(result) {
        if (result)
            this._view?.showSuccess();
        else
            this._view?.showError();
    }
}
//#region Constant
/**
 * 管理员列表存储键.
 */
GodModService.GodModAdminListStorageKey = "GOD_MOD_ADMIN_LIST__STORAGE_KEY";
/**
 * God Mod 请求运行 事件名前缀.
 */
GodModService.GodModRunRequestEventNamePrefix = "__GOD_MOD_RUN_REQUEST_EVENT_NAME__";
/**
 * God Mod 命令于服务端运行完成 事件名.
 */
GodModService.GodModCommandRunResultInServerEventName = "__GOD_MOD_COMMAND_RUN_RESULT_IN_SERVER__";
/**
 * God Mod 查询权限请求 事件名.
 */
GodModService.GodModQueryAuthorityReqEventName = "__GOD_MOD_QUERY_AUTHORITY_REQ_EVENT_NAME__";
/**
 * God Mod 查询权限回应 事件名.
 */
GodModService.GodModQueryAuthorityRespEventName = "__GOD_MOD_QUERY_AUTHORITY_RESP_EVENT_NAME__";
function addGMCommand(label, paramType, clientCmd = undefined, serverCmd = undefined, paramOption = undefined, group) {
    GodModService.getInstance().addCommand(label, paramType, clientCmd, serverCmd, paramOption, group);
}

export { GodCommandItem, GodModPanel, addGMCommand, GodModService as default };
