// mw-god-mod/dist/ui/param-input/GodModStringParamInput.js
import { ParamInputSizeY } from '../param-base/IGodModParamInput.js';
import Gtk from 'gtoolkit';
import { Lui, Component, TextField, Property } from 'mw-lynx-ui';

var Color = Lui.Asset.Color;
/**
 * GodModStringParamInput.
 *
 * ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟
 * ⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄
 * ⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄
 * ⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄
 * ⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
 * @author LviatYi
 * @font JetBrainsMono Nerd Font Mono https://github.com/ryanoasis/nerd-fonts/releases/download/v3.0.2/JetBrainsMono.zip
 * @fallbackFont Sarasa Mono SC https://github.com/be5invis/Sarasa-Gothic/releases/download/v0.41.6/sarasa-gothic-ttf-0.41.6.7z
 * @internal
 */
class GodModStringParamInput extends Component {
    //#region Lui Component
    static create(option) {
        let input = new GodModStringParamInput();
        input._input = TextField.create({
            label: "string",
            size: { x: 400, y: ParamInputSizeY },
            color: {
                primary: Color.Blue,
                secondary: Color.Blue200,
            },
            fontSize: 16,
            fontStyle: mw.UIFontGlyph.Light,
            corner: Property.Corner.Top,
        }).attach(input);
        Gtk.setUiSize(input.root, 400, ParamInputSizeY);
        return input;
    }
    ;
    //#endregion ⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠐⠒⠒⠒⠒⠚⠛⣿⡟⠄⠄⢠⠄⠄⠄⡄⠄⠄⣠⡶⠶⣶⠶⠶⠂⣠⣶⣶⠂⠄⣸⡿⠄⠄⢀⣿⠇⠄⣰⡿⣠⡾⠋⠄⣼⡟⠄⣠⡾⠋⣾⠏⠄⢰⣿⠁⠄⠄⣾⡏⠄⠠⠿⠿⠋⠠⠶⠶⠿⠶⠾⠋⠄⠽⠟⠄⠄⠄⠃⠄⠄⣼⣿⣤⡤⠤⠤⠤⠤⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄⠄
    //#region IGodModParamInputParametric
    getParam() {
        return this._input.text;
    }
    setParam(p) {
        this._input.setContent(p ?? "");
    }
    setValidator(validator) {
        this._input.setValidator(validator);
    }
    get validated() {
        return this._input.validated;
    }
    //#endregion
    //#region Init
    get onCommit() {
        return this._input.onCommit;
    }
    get onKeyUp() {
        return this._input.onKeyUp;
    }
    ;
}

export { GodModStringParamInput as default };
